test/fastify-integrations.spec.js

```diff
@@ -0,0 +1,85 @@
+'use strict'
+
+const { readFileSync } = require('fs')
+const path = require('path')
+const t = require('tap')
+const nock = require('nock')
+const Fastify = require('fastify')
+
+const jwks = require('./constants').jwks
+const buildGetJwks = require('../src/get-jwks')
+
+t.beforeEach((done) => {
+  nock.disableNetConnect()
+  done()
+})
+
+t.afterEach((done) => {
+  nock.cleanAll()
+  nock.enableNetConnect()
+  done()
+})
+
+t.test('integration tests', t => {
+  const domain = 'https://localhost/'
+
+  nock(domain).get('/.well-known/jwks.json').reply(200, jwks)
+
+  const jwk = jwks.keys[1]
+  const getJwks = buildGetJwks()
+  const customErrorMessages = {
+    badRequestErrorMessage: (err) => { console.log('Test Error: ', err.message); return err.message },
+    noAuthorizationInHeaderMessage: (err) => { console.log('Test Error: ', err.message); return err.message },
+    authorizationTokenExpiredMessage: (err) => { console.log('Test Error: ', err.message); return err.message },
+    authorizationTokenInvalid: (err) => { console.log('Test Error: ', err.message); return err.message }
+  }
+
+  const fastify = Fastify()
+  fastify.register(require('fastify-jwt'), {
+    ignoreExpiration: true,
+    decode: { complete: true },
+    secret: {
+      private: {
+        passphrase: 'mysecret',
+        key: readFileSync(path.join(__dirname, 'private.pem'), 'utf8')
+      },
+      public: async (request, token, callback) => {
+        const { header: { kid, alg }, payload: { iss } } = token
+        const publicKey = await getJwks.getPublicKey({ kid, domain: iss, alg })
+        callback(null, publicKey)
+      }
```

Comment:
this looks good now, and in fact a scenario I had not considered. I'm not sure it's a perfectly valid one to be honest. if you're using JWKS, you probably have the keys available so you don't need to ask them to an external service. In any case, we can keep this test but we should have one which is the usage you've have when integrating with an external service like cognito. the application doesn't know the private key, it only gets the signed token and uses the public key to verify it. in other words, I believe we need a test which doesn't use the secret.private property because we don't have that available. Instead, it uses the function version of the secret property or simply secret.public (not sure this works if you provide public but not private, please try it out) to provide a function which has the same implementation as your secret.public function above. then, you would not use fastify.jwt.sign to sign the token but you would sign it outside of the app using jsonwebtoken. I hope this is clear enough

Reply:
```
const jsonwebtoken = require('...')

const token = jsonwebtoken.sign({ payload }, privateKey, { issuer, keyid, algorithm })
```